src/main/scala/com/snowplowanalytics/refererparser/scala/Parser.scala

```diff
@@ -1,77 +1,117 @@
+
 // RefererParser Java impl
 import com.snowplowanalytics.refererparser.{Parser => JParser}
 import com.snowplowanalytics.refererparser.{Medium => JMedium}
 
 // Scala
-import scala.collection.JavaConversions._ 
+import scala.collection.JavaConversions._
+import scala.util.Try
+import com.typesafe.scalalogging.Logger
 
 /**
- * Enumeration for supported mediums.
- *
- * Replacement for Java version's Enum.
- */
+  * Enumeration for supported mediums.
+  *
+  * Replacement for Java version's Enum.
+  */
 object Medium extends Enumeration {
   type Medium = Value
 
-  val Unknown  = Value("unknown")
-  val Search   = Value("search")
+  val Unknown = Value("unknown")
+  val Search = Value("search")
   val Internal = Value("internal")
-  val Social   = Value("social")
-  val Email    = Value("email")
+  val Social = Value("social")
+  val Email = Value("email")
+  val Paid = Value("paid")
 
   /**
-   * Converts from our Java Medium Enum
-   * to our Scala Enumeration values above.
-   */
+    * Converts from our Java Medium Enum
+    * to our Scala Enumeration values above.
+    */
   def fromJava(medium: JMedium) =
     Medium.withName(medium.toString())
 }
 
 /**
- * Immutable case class to hold a referer.
- *
- * Replacement for Java version's POJO.
- */
+  * Immutable case class to hold a referer.
+  *
+  * Replacement for Java version's POJO.
+  */
 case class Referer(
-  medium: Medium.Medium,
-  source: Option[String],
-  term:   Option[String]
+    medium: Medium.Medium,
+    source: Option[String],
+    term: Option[String]
 )
 
 /**
- * Parser object - contains one-time initialization
- * of the YAML database of referers, and parse()
- * methods to generate a Referer object from a
- * referer URL.
- *
- * In Java this had to be instantiated as a class.
- */
+  * Parser object - contains one-time initialization
+  * of the YAML database of referers, and parse()
+  * methods to generate a Referer object from a
+  * referer URL.
+  *
+  * In Java this had to be instantiated as a class.
+  */
 object Parser {
 
   type MaybeReferer = Option[Referer]
 
-  private lazy val jp = new JParser()
+  val logger = Logger[Parser.type]
+
+  private lazy val jp: JParser = createParser
+
+  val createParser: JParser = {
+    import org.apache.http.client.fluent.Request
+    import org.apache.http.util.EntityUtils
+
+    val LatestReferrerUrl =
+      "https://s3-eu-west-1.amazonaws.com/snowplow-hosted-assets/third-party/referer-parser/referers-latest.yml"
+
+    lazy val latestParser =
+      for {
+        _ <- Try(logger.info(s"Trying to load referrers from $LatestReferrerUrl"))
+        response <- Try(Request.Get(LatestReferrerUrl).execute().returnResponse())
+        statusCode <- Try(response.getStatusLine.getStatusCode)
+        if statusCode == 200
+        inStream <- Try(new ByteArrayInputStream(EntityUtils.toByteArray(response.getEntity)))
+        parser <- Try(new JParser(inStream))
+      } yield parser
+
+    latestParser.transform({ p =>
```

Comment:
Doesn't this line force the evaluation of the `lazy` `latestParser`, thus making the `lazy` useless?